# Exploring the anomaly on 27th of February

## Throughout previous calculations of the funnel stages overview and later in visualization I found and anomaly -> spike in event Cart -> Purchase on 27th of February

1. Firstly, data is anonymous and I have no way to get the company/ state background if it's truly based on some event
2. Based on quick google seach there are no major holidays on February the 27th
3. Based on Google web searches from that time, not even Coronavirus was a greatly discussed topic at that time

Based on the spike on 27th only, it points out to a mistake in data, but let's see

## Based on my findings I decided that for futute visualizations (excluding median whitch accounts for extremes) exclude the data for the 27th of February

In [1]:
import pandas as pd

In [44]:
feb = pd.read_csv("2020-Feb.csv")

In [39]:
feb.tail(40)

event_time event_type  product_id          category_id  \
55318525  2020-02-29 23:59:41 UTC       view     5100876  2232732103101907535   
55318526  2020-02-29 23:59:41 UTC       view     4804055  2232732079706079299   
55318527  2020-02-29 23:59:42 UTC       view   100025092  2053013565194896323   
55318528  2020-02-29 23:59:42 UTC       view     3700995  2232732101063475749   
55318529  2020-02-29 23:59:42 UTC       view   100077498  2232732085846540487   
55318530  2020-02-29 23:59:44 UTC       view   100068488  2232732093077520756   
55318531  2020-02-29 23:59:45 UTC       view    15100270  2232732107413652135   
55318532  2020-02-29 23:59:45 UTC       view   100035134  2053013554658804075   
55318533  2020-02-29 23:59:45 UTC       view    17300555  2232732098446229999   
55318534  2020-02-29 23:59:46 UTC       view     1005100  2232732093077520756   
55318535  2020-02-29 23:59:46 UTC       view   100096952  2053013554155487563   
55318536  2020-02-29 23:59:46 UTC       view     2401325  2232732100769874463   
55318537  2020-02-29 23:59:47 UTC       view    30100074  2053013557670314521   
55318538  2020-02-29 23:59:47 UTC       view   100052511  2053013555757711783   
55318539  2020-02-29 23:59:47 UTC       view     1004767  2232732093077520756   
55318540  2020-02-29 23:59:47 UTC       view    32401995  2232732115777094520   
55318541  2020-02-29 23:59:48 UTC       view   100077498  2232732085846540487   
55318542  2020-02-29 23:59:48 UTC       view    42500010  2053013561495519987   
55318543  2020-02-29 23:59:49 UTC       view    28718192  2053013555321504139   
55318544  2020-02-29 23:59:49 UTC       view    18700066  2232732100056842769   
55318545  2020-02-29 23:59:49 UTC       view     6902464  2232732078733000751   
55318546  2020-02-29 23:59:49 UTC       view     5800792  2232732082390433922   
55318547  2020-02-29 23:59:49 UTC       view     1005160  2232732093077520756   
55318548  2020-02-29 23:59:50 UTC       cart     1004856  2232732093077520756   
55318549  2020-02-29 23:59:51 UTC       view   100032647  2053013558945383017   
55318550  2020-02-29 23:59:51 UTC       view    21410838  2232732082063278200   
55318551  2020-02-29 23:59:52 UTC       view   100064157  2053013557477376525   
55318552  2020-02-29 23:59:52 UTC       view     4100352  2232732098228126185   
55318553  2020-02-29 23:59:53 UTC       view    17303212  2232732098446229999   
55318554  2020-02-29 23:59:53 UTC       view   100129901  2053013553400512793   
55318555  2020-02-29 23:59:53 UTC       view    22700128  2232732091643068746   
55318556  2020-02-29 23:59:55 UTC       view    15100305  2053013557024391671   
55318557  2020-02-29 23:59:56 UTC       view    22700955  2232732091643068746   
55318558  2020-02-29 23:59:56 UTC       view     5100876  2232732103101907535   
55318559  2020-02-29 23:59:56 UTC       view     5801655  2232732082390433922   
55318560  2020-02-29 23:59:57 UTC       view   100118024  2232732099436085763   
55318561  2020-02-29 23:59:58 UTC       view    12200332  2053013553325015316   
55318562  2020-02-29 23:59:58 UTC       view     6902464  2232732078733000751   
55318563  2020-02-29 23:59:59 UTC       view   100103966  2053013560530830019   
55318564  2020-02-29 23:59:59 UTC       view    16000659  2053013556856619499   

                             category_code       brand    price    user_id  \
55318525                electronics.clocks     samsung   385.83  537926034   
55318526                     sport.bicycle       apple   193.75  614770853   
55318527             apparel.shoes.sandals         NaN    28.31  603050100   
55318528     appliances.environment.vacuum     shivaki   128.70  538268573   
55318529      construction.tools.generator      tamina   424.72  613391860   
55318530          construction.tools.light     samsung   293.13  517021211   
55318531                    apparel.shorts         NaN  1935.18  560360095   
55318532       electronics.audio.headphone          hp   308.63  622086728   

In [40]:
feb = feb.drop(columns=["category_id", "brand", "price","user_session","category_code", "product_id"])

In [41]:
feb.tail(20)

event_time event_type    user_id
55318545  2020-02-29 23:59:49 UTC       view  531574188
55318546  2020-02-29 23:59:49 UTC       view  532539344
55318547  2020-02-29 23:59:49 UTC       view  514753345
55318548  2020-02-29 23:59:50 UTC       cart  560677759
55318549  2020-02-29 23:59:51 UTC       view  518103713
55318550  2020-02-29 23:59:51 UTC       view  574420947
55318551  2020-02-29 23:59:52 UTC       view  553589144
55318552  2020-02-29 23:59:52 UTC       view  615486096
55318553  2020-02-29 23:59:53 UTC       view  616912131
55318554  2020-02-29 23:59:53 UTC       view  618035216
55318555  2020-02-29 23:59:53 UTC       view  599238796
55318556  2020-02-29 23:59:55 UTC       view  579864826
55318557  2020-02-29 23:59:56 UTC       view  540118001
55318558  2020-02-29 23:59:56 UTC       view  537926034
55318559  2020-02-29 23:59:56 UTC       view  532539344
55318560  2020-02-29 23:59:57 UTC       view  519011545
55318561  2020-02-29 23:59:58 UTC       view  540823940
55318562  2020-02-29 23:59:58 UTC       view  531574188
55318563  2020-02-29 23:59:59 UTC       view  614934914
55318564  2020-02-29 23:59:59 UTC       view  622090611

In [42]:
feb = feb.groupby("event_type")['user_id'].nunique().reset_index()

In [33]:
view_to_cart= feb.loc[feb['event_type'] == 'cart']['user_id'].values[0] / feb.loc[feb['event_type'] == 'view']['user_id'].values[0]
view_to_purchase= feb.loc[feb['event_type'] == 'purchase']['user_id'].values[0] / feb.loc[feb['event_type'] == 'view']['user_id'].values[0]
cart_to_purchase= feb.loc[feb['event_type'] == 'purchase']['user_id'].values[0] / feb.loc[feb['event_type'] == 'cart']['user_id'].values[0]

print(f"View to cart conversion rate: {view_to_cart:.2%}")
print(f"View to purchase conversion rate: {view_to_purchase:.2%}")
print(f"Cart to purchase conversion rate: {cart_to_purchase:.2%}")

View to cart conversion rate: 18.83%
View to purchase conversion rate: 9.33%
Cart to purchase conversion rate: 49.55%


In [45]:
# First, let's reload the raw data to preserve the date column for analysis
feb_raw = pd.read_csv("2020-Feb.csv")
feb_raw['event_time'] = pd.to_datetime(feb_raw['event_time'])
feb_raw['date'] = feb_raw['event_time'].dt.date

# Check the date range
print("Date range in data:")
print(f"Min date: {feb_raw['date'].min()}")
print(f"Max date: {feb_raw['date'].max()}")
print(f"\nEvent types: {feb_raw['event_type'].unique()}")
print(f"\nTotal rows: {len(feb_raw)}")

Date range in data:
Min date: 2020-02-01
Max date: 2020-02-29

Event types: ['view' 'cart' 'purchase']

Total rows: 55318565


In [47]:
# Calculate daily conversion rates
daily_stats = feb_raw.groupby(['date', 'event_type'])['user_id'].nunique().reset_index()
daily_stats['date'] = daily_stats['date'].astype(str)  # Convert dates to strings for proper indexing
daily_pivot = daily_stats.pivot(index='date', columns='event_type', values='user_id').fillna(0)
daily_pivot.columns.name = None

# Calculate conversion rates for each day
daily_pivot['view_to_cart'] = daily_pivot['cart'] / daily_pivot['view']
daily_pivot['view_to_purchase'] = daily_pivot['purchase'] / daily_pivot['view']
daily_pivot['cart_to_purchase'] = daily_pivot['purchase'] / daily_pivot['cart']

print("Daily conversion rates (sample around Feb 27):")
print(daily_pivot.loc['2020-02-24':'2020-02-29', ['view', 'cart', 'purchase', 'view_to_cart', 'cart_to_purchase', 'view_to_purchase']])
print("\n" + "="*80)
print(f"February 27th specifically:")
print(daily_pivot.loc['2020-02-27', ['view', 'cart', 'purchase', 'view_to_cart', 'cart_to_purchase', 'view_to_purchase']])

Daily conversion rates (sample around Feb 27):
              view   cart  purchase  view_to_cart  cart_to_purchase  \
date                                                                  
2020-02-24  315075  37635     19299      0.119448          0.512794   
2020-02-25  308032  37331     19296      0.121192          0.516889   
2020-02-26  313780  38159     19251      0.121611          0.504494   
2020-02-27   39288   2935     18707      0.074705          6.373765   
2020-02-28  180445  19856     18942      0.110039          0.953969   
2020-02-29  227838  23897     18017      0.104886          0.753944   

            view_to_purchase  
date                          
2020-02-24          0.061252  
2020-02-25          0.062643  
2020-02-26          0.061352  
2020-02-27          0.476150  
2020-02-28          0.104974  
2020-02-29          0.079078  

February 27th specifically:
view                39288.000000
cart                 2935.000000
purchase            18707.000000
view_to_

1. February 27th has a conversion rate 6.3 cart to purchase (other days only around 0.5-0.9)
2. What is more damming the views are 39 288 and purchases 18 707 on that day and cart only 2935 -> only 7.47% 
3. That is not constant with the rest of our data because that would mean 86,43% of buyers didn't use the cart -> again not constant with the rest of our data

# I decided that for future visualizations and calculations (excluding the median witch accounts for skewing and extreme data) I will exclude the date of 27th of February

In [49]:
# Let's investigate Feb 27 in detail
import datetime

feb27_date = datetime.date(2020, 2, 27)
feb26_date = datetime.date(2020, 2, 26)

print("=== Detailed Analysis for February 27th ===\n")

# Count raw events (not unique users) by event type
feb27_raw_counts = feb_raw[feb_raw['date'] == feb27_date].groupby('event_type').size()
print("Total events by type on Feb 27:")
print(feb27_raw_counts)
print()

# Compare with a normal day (Feb 26)
feb26_raw_counts = feb_raw[feb_raw['date'] == feb26_date].groupby('event_type').size()
print("Total events by type on Feb 26 (normal day):")
print(feb26_raw_counts)
print()

# Look at unique users by type
feb27_users = feb_raw[feb_raw['date'] == feb27_date].groupby('event_type')['user_id'].nunique()
print("Unique users by event type on Feb 27:")
print(feb27_users)
print()

# Check if there's something unusual about purchases on Feb 27
feb27_purchases = feb_raw[(feb_raw['date'] == feb27_date) & (feb_raw['event_type'] == 'purchase')]
print(f"Number of purchase events on Feb 27: {len(feb27_purchases)}")
print(f"Unique users with purchases on Feb 27: {feb27_purchases['user_id'].nunique()}")
print()

# Check if users made multiple purchases
feb27_user_purchase_counts = feb27_purchases['user_id'].value_counts()
print(f"Users making multiple purchases on Feb 27:")
print(f"Max purchases by one user: {feb27_user_purchase_counts.max()}")
print(f"Average purchases per user: {feb27_user_purchase_counts.mean():.2f}")
print(f"Users with 2+ purchases: {(feb27_user_purchase_counts > 1).sum()}")

=== Detailed Analysis for February 27th ===

Total events by type on Feb 27:
event_type
cart          6107
purchase     27499
view        163441
dtype: int64

Total events by type on Feb 26 (normal day):
event_type
cart          84213
purchase      27809
view        1800712
dtype: int64

Unique users by event type on Feb 27:
event_type
cart         2935
purchase    18707
view        39288
Name: user_id, dtype: int64

Number of purchase events on Feb 27: 27499
Unique users with purchases on Feb 27: 18707

Users making multiple purchases on Feb 27:
Max purchases by one user: 42
Average purchases per user: 1.47
Users with 2+ purchases: 4412
